omreader/src/t1/T1.scala

```diff
@@ -1,45 +1,54 @@
 // SPDX-License-Identifier: Apache-2.0
-// SPDX-FileCopyrightText: 2022 Jiuyang Liu <liu@jiuyang.me>
-
+// SPDX-FileCopyrightText: 2025 Yuhang Zeng <unlsycn@unlsycn.com>
 package org.chipsalliance.t1.omreader.t1
 
-import mainargs._
-import org.chipsalliance.t1.omreaderlib.t1.{T1 => Lib}
+import org.chipsalliance.t1.omreader.*
 
-object T1 {
-  implicit object PathRead extends TokensReader.Simple[os.Path] {
-    def shortName = "path"
-    def read(strs: Seq[String]): Either[String, os.Path] = Right(os.Path(strs.head, os.pwd))
-  }
-  @main
-  def vlen(@arg(name = "mlirbc-file") mlirbcFile: os.Path) =
-    println(new Lib(os.read.bytes(mlirbcFile)).vlen)
+import me.jiuyang.omlib.{*, given}
 
-  @main
-  def dlen(@arg(name = "mlirbc-file") mlirbcFile: os.Path) =
-    println(new Lib(os.read.bytes(mlirbcFile)).dlen)
+import upickle.default._
+import mainargs.*
 
-  @main
-  def instructions(@arg(name = "mlirbc-file") mlirbcFile: os.Path) =
-    new Lib(os.read.bytes(mlirbcFile)).instructions.foreach(println)
+class T1(mlirbc: Array[Byte]) extends OMReader(mlirbc, "T1_Class") with T1OMReader:
+  private val t1 = top.obj("om")
 
-  @main
-  def extensions(
-    @arg(name = "mlirbc-file") mlirbcFile: os.Path
-  ) =
-    println(new Lib(os.read.bytes(mlirbcFile)).extensions.mkString("_"))
+  given Writer[T1OM] = macroW
+  case class T1OM(
+    vlen:         OMValue,
+    dlen:         OMValue,
+    instructions: Seq[OMValue],
+    extensions:   Seq[OMValue],
+    march:        OMValue,
+    sram:         Seq[OMValue],
+    retime:       Seq[OMValue])
 
-  @main
-  def march(@arg(name = "mlirbc-file") mlirbcFile: os.Path) =
-    println(new Lib(os.read.bytes(mlirbcFile)).march)
+  val vlen = t1.obj("vlen")
 
-  @main
-  def sram(@arg(name = "mlirbc-file") mlirbcFile: os.Path) =
-    new Lib(os.read.bytes(mlirbcFile)).sram.foreach(s => println(upickle.default.write(s)))
+  val dlen = t1.obj("dlen")
+
+  val instructions = t1.obj("decoder").obj("instructions").list.toSeq
```

Comment:
I think in the open source part we don't require this metadata, but anyway let's keep it here, we will clean up them in the future.